Embed key into dict entry

hpatro opened the pull request:
This PR incorporates changes related to key embedding described in the https://github.com/redis/redis/issues/12216
With this change there will be no `key` pointer and embedded the key within the `dictEntry`. 1 byte is used for additional bookkeeping. Overall the saving would be 7 bytes.

Closes #394.

Key changes:

New dict entry type introduced, which is now used as an entry for the main dictionary:

```c
typedef struct {
    union {
        void *val;
        uint64_t u64;
        int64_t s64;
        double d;
    } v;
    struct dictEntry *next;  /* Next entry in the same hash bucket. */
    uint8_t key_header_size; /* offset into key_buf where the key is located at. */
    unsigned char key_buf[]; /* buffer with embedded key. */
} embeddedDictEntry;
```

One new function has been added to the dictType:

```c
size_t (*embedKey)(unsigned char *buf, size_t buf_len, const void *key, unsigned char *header_size);
```


Change is opt-in per dict type, hence sets, hashes and other types that are using dictionary are not impacted.
With this change main dictionary now owns the data, so copy on insert in dbAdd is no longer needed.

### Benchmarking results

TLDR; Around 9-10% memory usage reduction in overall memory usage for scenario with key of 16 bytes and value of 8 bytes and 16 bytes. The throughput per second varies but is similar or greater in most of the run(s) with the changes against unstable (ae2d421).

* Performed on a Amazon EC2 c5.4xlarge instance
* Server setup (maximum memory allocated is 100 MB)
```
src/valkey-server --save "" --daemonize yes --maxmemory 100m --enable-debug-command local --port 6379
```
* GET command used
```
src/valkey-benchmark  -t get  -n 1000000  -r 10000000
 ```
 * SET command used
 ```
 src/valkey-benchmark  -t get  -n 1000000  -r 10000000 -d 16
 ```

#### SET performance

  |  Throughput per sec  |  Number of keys  |   Used memory (bytes)  |   Throughput per sec   |  Number of keys  |   Used memory (bytes)  |  Throughput per sec  |  Number of keys  |  Used memory (bytes) 
-- | -- | -- | -- | -- | -- | -- | -- | -- | --
Key Embedding |   |   |   |   |   |   |   |   |  
SET (d = 8, n= 1M, r=10M) | 108448.11 | 951737 | 78140504 | 107135.21 | 951660 | 78135144 | 108026.36 | 951616 | 78132160
SET (d = 16, n= 1M, r=10M) | 106974.76 | 951677 | 85750152 | 107342.21 | 951611 | 85745056 | 107215.62 | 951757 | 85754792
  |   |   |   |   |   |   |   |   |  
Unstable (ae2d421) |   |   |   |   |   |   |   |   |  
SET (d = 8, n= 1M, r=10M) | 106929 | 951556 | 85715672 | 107009.09 | 951671 | 85725056 | 106484.94 | 951864 | 85740680
SET (d = 16, n= 1M, r=10M) | 104036.62 | 951541 | 93226392 | 105965.88 | 951710 | 93341232 | 104123.28 | 951627 | 93333560



#### GET performance


Operations |  Number of Keys  |  Used memory  | Throughput per sec | Throughput per sec | Throughput per sec 
-- | -- | -- | -- | -- | --
  |   |   |   |   |  
Key Embedding |   |   |   |   |  
GET (d = 8, n= 1M, r=10M) | 951580 | 78079272 | 105797.72 | 106168.39 | 107135.21
GET (d = 16, n= 1M, r=10M) | 951764 | 85731504 | 105797.72 | 106416.94 | 106202.2
  |   |   |   |   |  
Unstable (ae2d421) |   |   |   |   |  
GET (d = 8, n= 1M, r=10M) | 951213 | 85662008 | 106145.85 | 105685.91 | 105741.78
GET (d = 16, n= 1M, r=10M) | 951646 | 93260288 | 106723.59 | 106089.55 | 106349.04

hpatro committed:
Embed key into dict entry

Signed-off-by: Harkrishn Patro <harkrisp@amazon.com>

hpatro commented:
@valkey-io/core-team Could you provide some clarity on this ?

hpatro commented:
My thought captured under https://github.com/valkey-io/valkey/issues/394#issuecomment-2128139503 

*We've made changes in the past which can benefit the users in a short term and have redone the implementation in the following major/minor version. key embedding in dictEntry is a pretty small change and we can easily get rid of it with dictEntry removal. I feel the small gain is worth it with this minimal change (7 bytes) for Valkey 8.0 and invest on the kvpair object structure with open addressing in 9.0*

zuiderkwast commented:
Yeah, I've thought about this many times. The complexity is not too bad. The technique of embedding an sds string can later be reused in other places.

Although the embedding is abstracted, it only every makes sense for sds strings. That's fine though. Dict doesn't include "sds.h" so it's decoupled.

I'm in favor. (I'll add some review comments later, just minor things.)

The reason I've been skeptical before is that I'd rather like that we invest in embedding key in robj, since that'd be beneficial in the future redesign (#169), but since this PR is already ready and the robj work is not started, I think we can merge this for Valkey 8.

bbarani commented:
@hpatro @madolson Is this change targeted for Valkey 8? If so, can we add it to Valkey 8 project [board](https://github.com/orgs/valkey-io/projects/2)?

zuiderkwast commented on src/defrag.c:
```diff
@@ -724,7 +740,7 @@ void defragKey(defragCtx *ctx, dictEntry *de) {
 void defragScanCallback(void *privdata, const dictEntry *de) {
     long long hits_before = server.stat_active_defrag_hits;
-    defragKey((defragCtx *)privdata, (dictEntry *)de);
+    defragItem((defragCtx *)privdata, (dictEntry *)de);
```
Item?

We refer to a key-value pair incl TTL etc simply as a "key". We're talking about key types (which actually refers to the value).

I don't think we should rename this to "item" at this point.

madolson commented:
I was waiting on performance numbers from Hari before officially adding it.

bbarani commented:
> I was waiting on performance numbers from Hari before officially adding it.

@hpatro Do you have performance numbers? Can you please add it to this issue to move forward with next steps?

PingXie commented on src/dict.c:
```diff
@@ -124,6 +135,7 @@ uint64_t dictGenCaseHashFunction(const unsigned char *buf, size_t len) {
 #define ENTRY_PTR_MASK 7     /* 111 */
 #define ENTRY_PTR_NORMAL 0   /* 000 */
 #define ENTRY_PTR_NO_VALUE 2 /* 010 */
+#define ENTRY_PTR_EMBEDDED 4 /* 100 */
```
[Readability suggestion]

looks like we are missing `1`. Is it currently used a key flag in `entryIsKey`? If so, it would be good to add it too.

zuiderkwast replied:
@PingXie yes, `xx1` means "entry is key". Add a comment for it, or do you have another suggestion?

PingXie replied:
maybe add a define? magic values are always hard to keep track of ...

```
#define ENTRY_PTR_IS_KEY 1
```

hpatro replied:
Yeah, will add it. It's the trick @zuiderkwast used for entry is key as sds data pointer address is always an odd address because the metadata struct are all odd in size.

zuiderkwast replied:
The `ENTRY_PTR_NORMAL` and the others here are taken from the pointer's last three bits, i.e. it's masked with `ENTRY_PTR_MASK`.

The pointer-is-key flag is just one bit. What's the point of having a macro defined to 1? Do you also want to define a 1-bit mask for it too which explains which bits to use for this judgement?

    #define ENTRY_PTR_IS_KEY_MASK 1
    #define ENTRY_PTR_IS_KEY 1

I think that's excessive. These macros are only used in the next inline functions anyway. The inline functions make up the internal abstraction for these encoded pointers.

```C
static inline int entryIsKey(const dictEntry *de) {
    return (uintptr_t)(void *)de & 1;
}

static inline int entryIsNormal(const dictEntry *de) {
    return ((uintptr_t)(void *)de & ENTRY_PTR_MASK) == ENTRY_PTR_NORMAL;
}
```

@PingXie if you want an opaque abstraction of this, then I think the only way is to put this part in a separate file and only expose certain functions to it. Is it worth it?

zuiderkwast replied:
I'd say we shouldn't change this abstraction just because we're using one of the unused values here. What we could do is to add a comment like this, just to show that any pattern with the last bit = 1 are taken.

```suggestion
#define ENTRY_PTR_EMBEDDED 4 /* 100 */
/*      ENTRY_PTR_IS_KEY        xx1 */
```

PingXie commented on src/dict.c:
```diff
@@ -74,6 +74,17 @@ struct dictEntry {
     struct dictEntry *next; /* Next entry in the same hash bucket. */
 };
 
+typedef struct {
+    union {
+        void *val;
+        uint64_t u64;
+        int64_t s64;
+        double d;
+    } v;
+    struct dictEntry *next; /* Next entry in the same hash bucket. */
+    unsigned char data[];
+} embeddedDictEntry;
+
```
[Readability suggestion] I think we can get rid of the magic constant, `ENTRY_METADATA_BYTES`. Also I think `key_buf` conveys the purpose of this memory region clearer than `data`

```suggestion
typedef struct {
    union {
        void *val;
        uint64_t u64;
        int64_t s64;
        double d;
    } v;
    struct dictEntry *next; /* Next entry in the same hash bucket. */
    unsigned char key_offset; /* offset into key_buf where the key is located at. */
    char key_buf[];
} embeddedDictEntry;
```

madolson replied:
I was going to ask, what is ENTRY_METADATA_BYTES even for?

madolson replied:
Just a note, the `unsigned char key_offset` will get padded to 8 bytes in a sizeof calculation, so we need to subtract off 7 when we do the math during the zmalloc otherwise we will over allocate memory. Either that or we need to allocate a normal dictEntry and just cast it to an embeddedDictEntry.

It also feels like we are tightly coupling the naming to the sds, which has a concept of having the start of the the key part way into the buffer with a header in the prefix. I would rather just call it key_header_size.

PingXie replied:
> the unsigned char key_offset will get padded to 8 bytes

Good point. We could pack the struct ourselves and `static_assert` on `v` and `next` being aligned

```c
typedef struct __attribute__((packed)) {
    union {
        void *val;
        uint64_t u64;
        int64_t s64;
        double d;
    } v;
    struct dictEntry *next; /* Next entry in the same hash bucket. */
    unsigned char key_offset; /* offset into key_buf where the key is located at. */
    char key_buf[];
} embeddedDictEntry;
```

> I would rather just call it key_header_size.

SGTM

hpatro replied:
Vitaly and I had also planned to abstract lot of these metadata values out to individual variables for better readability.

hpatro replied:
Getting the following error with the packed struct. checking how to avoid it.

```
dict.c:930:37: error: taking address of packed member of ‘struct <anonymous>’ may result in an unaligned pointer value [-Werror=address-of-packed-member]
  930 |     if (entryIsEmbedded(de)) return &decodeEmbeddedEntry(de)->next;
      |                                     ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
cc1: all warnings being treated as errors
```

madolson replied:
@hpatro I think you can do this? We only care about the last data being packed, the rest of the data will be properly aligned.
```
typedef struct {
    union {
        void *val;
        uint64_t u64;
        int64_t s64;
        double d;
    } v;
    struct dictEntry *next; /* Next entry in the same hash bucket. */
    typedef struct __attribute__((packed)) {
        unsigned char key_offset; /* offset into key_buf where the key is located at. */
        char key_buf[];
    }
} embeddedDictEntry;
```

hpatro replied:
@madolson `sizeof` reported is still 24 bytes with the above. So, proceeding with your first suggestion. If anyone else has a better idea, will revisit this. Introduced a [compactSizeEmbeddedDictEntry](https://github.com/valkey-io/valkey/pull/541/commits/029917f56ae71a822d1f10889544936bc530dad6) method to find the size for allocation.

PingXie commented on src/server.c:
```diff
@@ -290,6 +290,18 @@ int dictSdsKeyCompare(dict *d, const void *key1, const void *key2) {
 }
 
+size_t dictSdsKeyLen(const void *key) { 
+    return sdsAllocSize((sds)key);
```
`sdsAllocSize` can return more than we need. `sdslen((sds)key)) + sdsHrdSize((sds)key) + 1` gives the tightest range.

PingXie commented on src/server.c:
```diff
@@ -290,6 +290,18 @@ int dictSdsKeyCompare(dict *d, const void *key1, const void *key2) {
+    return sdsAllocSize((sds)key);
+}
+
+size_t dictSdsKeyToBytes(unsigned char *buf, const void *key, unsigned char *header_size) {
```
[Readability suggestion] This function essentially replicate an sds string to a caller provided buffer. The only difference from `sdsdup` is that the caller allocates the memory. I think this logic belongs to sds.c.

madolson replied:
+1. It also solves some other oddities since functions like sdsHdrSize aren't exposed today, and probably shouldn't.

hpatro replied:
The problem was dict.c/dict.h can't access sds methods. So, server.c acts like a sort of middleware. I think I can keep this method as a facade and pullout the actual logic to an sds method.

PingXie commented on src/server.c:
```diff
@@ -290,6 +290,18 @@ int dictSdsKeyCompare(dict *d, const void *key1, const void *key2) {
+
+size_t dictSdsKeyToBytes(unsigned char *buf, const void *key, unsigned char *header_size) {
+    sds keysds = (sds)key;
+    size_t n_bytes = sdsAllocSize(keysds);
```
[Readability suggestion] This size computation logic needs to match `dictSdsKeyToBytes` to avoid buffer overrun bugs.

```suggestion
    size_t n_bytes = dictSdsKeyLen(keysds);
```

 Or better yet, let the caller provide the buffer length and we can assert on it being larger than `dictSdsKeyLen(key)`
 
 ```c
 size_t dictSdsKeyCopy(unsigned char *buf, size_t buf_len, const void *key, unsigned char *key_offset)
 ```

PingXie commented on src/server.c:
```diff
@@ -474,9 +486,12 @@ dictType dbDictType = {
     dictObjectDestructor, /* val destructor */
     dictResizeAllowed,    /* allow to resize */
+    .keyLen = dictSdsKeyLen,
+    .keyToBytes = dictSdsKeyToBytes,
```
```suggestion
    .keyCopy = dictSdsKeyCopy,
```

madolson replied:
In the context of a dictionary, I don't think `keyCopy` makes sense as an API name here. I would assume keyCopy creates a copy of the Key given something like a dictionary entry. I think the same is also true about the keyLen API. I would almost prefer to add embedded, something like: 
```suggestion
    .embededKeyLen = whatever,
    .embedKey = whatever,
```

PingXie replied:
Yeah `embedKey` is better but maybe we can remove `embedKeyLen`? I am thinking that we can have 
`dt->embedKey(NULL, 0, key, &entry->key_header_size)` return both the header size and the required buffer size, similar to `writePingExt`, https://github.com/valkey-io/valkey/blob/unstable/src/cluster_legacy.c#L3680

madolson replied:
I like that idea, having one API seems like a good approach.

PingXie commented on src/dict.c:
```diff
@@ -151,6 +168,15 @@ static inline dictEntry *createEntryNoValue(void *key, dictEntry *next) {
+    size_t keylen = dt->keyLen(key);
+    embeddedDictEntry *entry = zmalloc(sizeof(*entry) + keylen + ENTRY_METADATA_BYTES);
+    size_t bytes_written = dt->keyToBytes(entry->data + ENTRY_METADATA_BYTES, key, &entry->data[0]);
+    assert(bytes_written == keylen);
```
[Readability suggestion] With my other suggestions,

```suggestion
    dt->keyCopy(entry->key_buf, keylen, key, &entry->key_offset);
```

PingXie commented on src/dict.c:
```diff
@@ -161,15 +187,24 @@ static inline void *decodeMaskedPtr(const dictEntry *de) {
 
+static inline void *getEmbeddedKey(const dictEntry *de) {
+    embeddedDictEntry *entry = (embeddedDictEntry *)decodeMaskedPtr(de);
+    return &entry->data[ENTRY_METADATA_BYTES + entry->data[0]];
```
```suggestion
    return &entry->key_buf[entry->key_offset];
```

PingXie commented on src/dict.c:
```diff
@@ -161,15 +187,24 @@ static inline void *decodeMaskedPtr(const dictEntry *de) {
 /* Returns 1 if the entry has a value field and 0 otherwise. */
 static inline int entryHasValue(const dictEntry *de) {
-    return entryIsNormal(de);
+    return entryIsNormal(de) || entryIsEmbedded(de);
```
```suggestion
    return !entryIsNoValue(de);
```

PingXie commented on src/dict.c:
```diff
@@ -656,6 +693,9 @@ void dictFreeUnlinkedEntry(dict *d, dictEntry *he) {
     dictFreeKey(d, he);
     dictFreeVal(d, he);
+    /* Clear the embedded data */
+    if (entryIsEmbedded(he)) zfree(decodeEmbeddedEntry(he)->data);
```
This doesn't look right. `->data` is an embedded field in `embeddedDictEntry` and shouldn't be freed separately.

PingXie commented on src/dict.h:
```diff
@@ -83,6 +86,7 @@ typedef struct dictType {
     unsigned int keys_are_odd : 1;
     /* TODO: Add a 'keys_are_even' flag and use a similar optimization if that
```
curious - what is the 'keys_are_even' flag?

zuiderkwast replied:
@PingXie It doesn't exist. When I added `keys_are_odd`, Oran wanted me to add a TODO for this, so I added this comment, but it's not possible to use the same trick as we do for odd keys (as we do now, using LSB=1 to say it's a key instead of a dictEntry pointer).

madolson replied:
I always thought we should remove the keys_are_even TODO, but 🤷

zuiderkwast replied:
Me too. Let's remove the todo. :)

PingXie commented on src/object.c:
```diff
@@ -1559,7 +1559,7 @@ NULL
         }
         size_t usage = objectComputeSize(c->argv[2], dictGetVal(de), samples, c->db->id);
-        usage += sdsZmallocSize(dictGetKey(de));
+        usage += sdsAllocSize((sds)dictGetKey(de));
```
should we dedup these two functions? `sdsZmallocSize` seems to be the pattern in object.c

hpatro replied:
sdsZmallocSize won't work as this part of a bigger allocation. I've added `zmalloc_size` for the entire allocation and removed the `sdsAllocSize` and `dictEntryMemUsage`.

PingXie commented on src/dict.c:
```diff
@@ -509,6 +544,8 @@ dictEntry *dictInsertAtPosition(dict *d, void *key, void *position) {
             entry = createEntryNoValue(key, *bucket);
         }
+    } else if (d->type->embedded_entry) {
+        entry = createEmbeddedEntry(key, *bucket, d->type);
```
The way `dictEntry` is used has no type safety. Admittedly, this is not a new issue but the addition of `embeddedDictEntry` is making it worse. The following code path looks problematic to me. can you please double check?

`setGenericComamd() -> setKey() -> dbAdd() -> dbAddInsternal() -> kvstoreDictSetKey() -> dictSetKey()`

I don't see`dictSetKey` getting patched to handle this new embedded dict entry.

I would suggest making `dictEntry` an opaque struct next and force every function go through an inline accessor function/macro. I think this is the only certain way to ensure we don't accidentally use the wrong data type.

madolson replied:
Isn't the dictEntry already opaque?

PingXie replied:
not in dict.c.

madolson replied:
~~I guess I don't agree it should be opaque in dict.c. It seems like a very small number of actual touch points we actually have to make sure are correct. We could even reduce the number. I'm not sure making it opaque will help all that much.~~

Discussed offline, Ping has a separate proposal for adding guardrails that he will publish.

PingXie reviewed (dismissed):
This is a risky PR IMO. I am concerned about the mixed use of various dict entry types while using `dictEntry` as the "universal" pointer. I don't think it is feasible for me to examine every use of `dictEntry`, which should've been compiler's job. I am happy to help out on the refactoring if needed. Let me know.

hpatro commented:
Thanks for the review @PingXie and @zuiderkwast . I will shortly address them.

@madolson I've posted the benchmarking results on the top comment.

madolson commented on src/dict.c:
```diff
@@ -151,6 +168,15 @@ static inline dictEntry *createEntryNoValue(void *key, dictEntry *next) {
+    size_t bytes_written = dt->keyToBytes(entry->data + ENTRY_METADATA_BYTES, key, &entry->data[0]);
+    assert(bytes_written == keylen);
+    entry->next = next;
+    return (dictEntry *)(void *)((uintptr_t)(void *)entry | ENTRY_PTR_EMBEDDED);
```
```suggestion
    return encodeMaskedPtr(entry, ENTRY_PTR_EMBEDDED);
```

madolson reviewed (commented):
I still think overall this is a good iterative improvement before we can make more changes later. I would like to see us embed the key into the robj in the future, so this has a lot of the same primitives as that.

hpatro commented:
@zuiderkwast @PingXie @madolson @valkey-io/core-team If we all are aligned on accepting this change in for 8.0, I will look into addressing the comments and polishing the PR further. Let me know.

hwware commented:
it looks like used memory decreases 10% without changing on qps.

madolson commented:
@hpatro AFAIK most of the folks are inclined to accept it for 8, so I would ask you to follow up if you can. I'll throw it on our Monday agenda as well to make sure we close on it quickly if we can.

PingXie commented:
> @zuiderkwast @PingXie @madolson @valkey-io/core-team If we all are aligned on accepting this change in for 8.0, I will look into addressing the comments and polishing the PR further. Let me know.

@hpatro, I like the idea! :-) The only reason I marked this PR as "changes required" is because of the amount of type casting in dict.c (and I understand that it didn't start with this PR). This is a great improvement to be had for Valkey 8.0 but I do think we need to make some potentially painful changes to get dict.c back in shape. If you don't mind, I would love to get my hands dirty and help out with the refactoring too.

hpatro commented:
> > @zuiderkwast @PingXie @madolson @valkey-io/core-team If we all are aligned on accepting this change in for 8.0, I will look into addressing the comments and polishing the PR further. Let me know.
> 
> 
> 
> @hpatro, I like the idea! :-) The only reason I marked this PR as "changes required" is because of the amount of type casting in dict.c (and I understand that it didn't start with this PR). This is a great improvement to be had for Valkey 8.0 but I do think we need to make some potentially painful changes to get dict.c back in shape. If you don't mind, I would love to get my hands dirty and help out with the refactoring too.

Thanks for the help @PingXie. Let me publish the changes which I have done tomorrow and we can go from there. Do we want to do further refactoring as a follow up PR or along with this?

hpatro committed:
Address feedback pass 1

Signed-off-by: Harkrishn Patro <harkrisp@amazon.com>

PingXie commented:
Thanks @hpatro.

>Do we want to do further refactoring as a follow up PR or along with this?

I am inclined to reduce as much type casting as possible in this PR and reason being that I don't trust myself doing the compiler job (of type checking the data access). I am concerned about the **potential** memory corruption issues. Will dedicate some time later this week for a deep-dive.

hpatro committed:
Address feedback pass 2

Signed-off-by: Harkrishn Patro <harkrisp@amazon.com>

hpatro committed:
Address feedback pass 3

Signed-off-by: Harkrishn Patro <harkrisp@amazon.com>

hpatro commented:
Couple of things which remain 

* https://github.com/valkey-io/valkey/pull/541#discussion_r1649820995 - Abstract out metadata field to independent variable, it cause issue with build - https://github.com/hpatro/valkey/actions/runs/9701100786/job/26773976709 (I'm trying out few things, let me know if anyone is aware of any solution to avoid this.)
```
dict.c: In function ‘dictGetNextRef’:
dict.c:930:37: error: taking address of packed member of ‘struct <anonymous>’ may result in an unaligned pointer value [-Werror=address-of-packed-member]
  930 |     if (entryIsEmbedded(de)) return &decodeEmbeddedEntry(de)->next;
      |                                     ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
cc1: all warnings being treated as errors
``` 

* https://github.com/valkey-io/valkey/pull/541#discussion_r1649829240 - @PingXie is looking to submit a proposal.

hpatro committed:
Compact allocation for embedded dict entry

Signed-off-by: Harkrishn Patro <harkrisp@amazon.com>

hpatro committed:
Update code comment

Signed-off-by: Harkrishn Patro <harkrisp@amazon.com>

hpatro commented:
@madolson I can't see any other builds apart from DCO check. Are we throttled/out of credits?

madolson commented on src/dict.c:
```diff
@@ -48,6 +49,10 @@
 #include "monotonic.h"
 
+#ifndef static_assert
+#define static_assert(expr, lit) extern char __static_assert_failure[(expr) ? 1 : -1]
```
We moved to c11, so we just have static asserts now:
```suggestion
#define static_assert(expr, lit) _Static_assert(expr, lit)
```

madolson commented:
> @madolson I can't see any other builds apart from DCO check. Are we throttled/out of credits?

You have a merge conflict..

zuiderkwast commented on src/dict.c:
```diff
@@ -74,6 +79,30 @@ struct dictEntry {
+/* The minimum amount of bytes required for embedded dict entry. */
+static inline size_t compactSizeEmbeddedDictEntry(void) {
+    return offsetof(embeddedDictEntry, key_buf);
+}
```
That's smart!

zuiderkwast commented on src/dict.c:
```diff
@@ -161,15 +189,40 @@ static inline void *decodeMaskedPtr(const dictEntry *de) {
 /* Returns 1 if the entry has a value field and 0 otherwise. */
 static inline int entryHasValue(const dictEntry *de) {
-    return entryIsNormal(de);
+    return !entryIsNoValue(de);
```
I believe we need to exclude entryIsKey too.

```suggestion
    return !entryIsNoValue(de) && !entryIsKey(de);
```

hpatro replied:
I think I got it right initially, putting it back to

```
return entryIsNormal(de) || entryIsEmbedded(de);
```

madolson commented on src/dict.c:
```diff
@@ -74,6 +79,30 @@ struct dictEntry {
+    } v;
+    struct dictEntry *next; /* Next entry in the same hash bucket. */
+    uint8_t key_header_size; /* offset into key_buf where the key is located at. */
+    unsigned char key_buf[]; /* buffer to embed the key. */
```
```suggestion
    unsigned char key_buf[]; /* buffer to embedded key. */
```

hpatro replied:
Isn't it correct `buffer to embed the key.` ?

madolson replied:
`buffer with embedded key`? The phrasing of `buffer to embed the key` implies that it will later be embeded with a key, but it's always embedded with the key.

hpatro committed:
Address feedback

Signed-off-by: Harkrishn Patro <harkrisp@amazon.com>

madolson commented on src/dict.h:
```diff
@@ -67,6 +67,8 @@ typedef struct dictType {
     /* Allow a dict to carry extra caller-defined metadata. The
      * extra memory is initialized to 0 when a dict is allocated. */
     size_t (*dictMetadataBytes)(dict *d);
+    size_t (*embedKey)(unsigned char *buf, size_t buf_len, const void *key, unsigned char *header_size);
```
```suggestion
    /* Function for copying a given key into a buffer of buf_len. Also used for
     * computing the length of the key + header when buf is NULL. */
    size_t (*embedKey)(unsigned char *buf, size_t buf_len, const void *key, unsigned char *header_size);
```

hpatro committed:
Merge branch 'unstable' into valkey-key-embedding

Signed-off-by: Harkrishn Patro <harkrisp@amazon.com>

hpatro committed:
Add code comment

Signed-off-by: Harkrishn Patro <harkrisp@amazon.com>

zuiderkwast commented on src/dict.c:
```diff
@@ -509,6 +562,8 @@ dictEntry *dictInsertAtPosition(dict *d, void *key, void *position) {
             entry = createEntryNoValue(key, *bucket);
         }
+    } else if (d->type->embedded_entry) {
+        entry = createEmbeddedEntry(key, *bucket, d->type);
```
If the dict type has no key dup callback, then `dictAddRaw()` used to take ownership of the key. This semantics is changed here if embedded key is enabled for the dict type. Then we copy the key instead. This means that the caller needs to free the key after calling `dictAddRaw()`.

That's why you needed to insert an `sdsfree()` call in rdb.c, I believe.

I suppose it affects `kvstoreDictAddRaw()` too, since it's calling dictAddRaw, and in turn, all functions calling kvstoreDictAddRaw, which are dbAddInternal, dbAddRDBLoad and setExpire.

I don't know if that's all. Maybe there are more...

pubsubSubscribeChannel calls kvstoreDictAddRaw too, but in this case, the dicts aren't using embedded entries (right?), so I suppose kvstoreDictAddRaw (and dictAddRaw) still takes ownership of the key in this case. (Maybe we should use embedded key for pubsub channels dicts too, so we can document that kvstoreDictAddRaw always copies the key?)

Short story: It becomes a mess. :)

We should at least document this behaviour at dictAdd, dictAddRaw and kvstoreDictAddRaw.

hpatro replied:
There are multiple levels of API to add data to the dictionary and would be lot of duplication of code comment. Should we add more documentation at the dictType level itself?

madolson replied:
Duplication of documentation is okay, since that is what people will actually refer to. I think we should update the documentation in the three places viktor mentioned.

zuiderkwast replied:
Sure, we can add some docs for dictType too, but I'm not sure what we should write there.

The caller of e.g. `pubsubSubscribeChannel()` may not know (and shouldn't need to know) that it using dictAddRaw and that the dictType has no val dup function, so it's better to document each function separately.

hpatro replied:
Let me know if the wording looks correct.

https://github.com/valkey-io/valkey/pull/541/commits/ce76a16ef40e81e7a0c08dd902f7961dbcaa3da0

zuiderkwast commented on src/rdb.c:
```diff
@@ -3261,6 +3261,9 @@ int rdbLoadRioWithLoadingCtx(rio *rdb, int rdbflags, rdbSaveInfo *rsi, rdbLoadin
             moduleNotifyKeyspaceEvent(NOTIFY_LOADED, "loaded", &keyobj, db->id);
+
+            /* Release key (sds), dictEntry stores a copy of it in embedded data */
+            sdsfree(key);
```
This is because of the implicitly changed semantics of dbAddRDBLoad?

```C
/* This is a special version of dbAdd() that is used only when loading
 * keys from the RDB file: the key is passed as an SDS string that is
 * retained by the function (and not freed by the caller).
 *
 * Moreover this function will not abort if the key is already busy, to
 * give more control to the caller, nor will signal the key as ready
 * since it is not useful in this context.
 *
 * The function returns 1 if the key was added to the database, taking
 * ownership of the SDS string, otherwise 0 is returned, and is up to the
 * caller to free the SDS string. */
int dbAddRDBLoad(serverDb *db, sds key, robj *val) {
    int slot = getKeySlot(key);
    dictEntry *de = kvstoreDictAddRaw(db->keys, slot, key, NULL);
    if (de == NULL) return 0;
    initObjectLRUOrLFU(val);
    kvstoreDictSetVal(db->keys, slot, de, val);
    return 1;
}
```

We need to update the documentation comment for that function. It no longer retains the key. Now it makes a copy without taking ownership of the key.

zuiderkwast reviewed (commented):
I found some difference in how dictAddRaw is copying the key vs taking ownership of the key. I guess the failed address sanitizer runs are related to this. I added some comments with some details.

zuiderkwast commented on src/dict.c:
```diff
@@ -76,6 +81,30 @@ struct dictEntry {
+} embeddedDictEntry;
+
+/* Validation and helper for `embeddedDictEntry` */
+static_assert(sizeof(embeddedDictEntry) == 24, "unexpected total size of embeddedDictEntry");
+static_assert(offsetof(embeddedDictEntry, v) == 0, "unexpected field offset");
+static_assert(offsetof(embeddedDictEntry, next) == 8, "unexpected field offset");
+static_assert(offsetof(embeddedDictEntry, key_header_size) == 16, "unexpected field offset");
+static_assert(offsetof(embeddedDictEntry, key_buf) == 17, "unexpected field offset");
```
... And this is why the 32-bit build fails. (A pointer is 32 bits here.)

Do we need these static asserts?

madolson replied:
Since we can't pack it, it would be nice to know that we aren't accidentally regressing the size. We really just want to make sure key_buf is placed right. It should be 2*sizeof(ptr) + 1.

zuiderkwast replied:
> We really just want to make sure key_buf is placed right.

Fair enough.


> It should be 2*sizeof(ptr) + 1.

Almost. The first element, the value, is a `union { void *val; uint64_t u64; int64_t s64; double d; }`.

I think `sizeof(ptr) + 9` is a better formula.

hpatro committed:
Accurate dictEntry memory usage calculation

Signed-off-by: Harkrishn Patro <harkrisp@amazon.com>

hpatro committed:
Format

Signed-off-by: Harkrishn Patro <harkrisp@amazon.com>

hpatro commented:
> I found some difference in how dictAddRaw is copying the key vs taking ownership of the key. I guess the failed address sanitizer runs are related to this. I added some comments with some details.

The failed address sanitizer were because of me doing a `zmalloc_size` on a encoded dictEntry pointer directly. Tried addressing it with this https://github.com/valkey-io/valkey/pull/541/commits/4d30109c3f8de79869fb80b0f50983f39f801ca4

hpatro committed:
Address feedback

Signed-off-by: Harkrishn Patro <harkrisp@amazon.com>

hpatro committed:
Try to make 32 bit compilation happy

Signed-off-by: Harkrishn Patro <harkrisp@amazon.com>

hpatro committed:
Remove total size check

Signed-off-by: Harkrishn Patro <harkrisp@amazon.com>

hpatro committed:
clang format

Signed-off-by: Harkrishn Patro <harkrisp@amazon.com>

PingXie commented:
> Couple of things which remain
> 
> * [Embed key into dict entry #541 (comment)](https://github.com/valkey-io/valkey/pull/541#discussion_r1649820995) - Abstract out metadata field to independent variable and make the struct packed, it cause issue with build - https://github.com/hpatro/valkey/actions/runs/9701100786/job/26773976709 (I'm trying out few things, let me know if anyone is aware of any solution to avoid this.)
> 
> ```
> dict.c: In function ‘dictGetNextRef’:
> dict.c:930:37: error: taking address of packed member of ‘struct <anonymous>’ may result in an unaligned pointer value [-Werror=address-of-packed-member]
>   930 |     if (entryIsEmbedded(de)) return &decodeEmbeddedEntry(de)->next;
>       |                                     ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
> cc1: all warnings being treated as errors
> ```
> 
> * [Embed key into dict entry #541 (comment)](https://github.com/valkey-io/valkey/pull/541#discussion_r1649829240) - @PingXie is looking to submit a proposal.

Yeah I did some experiments. It is actually not that big a [change](https://github.com/valkey-io/valkey/compare/unstable...PingXie:valkey:valkey-key-embedding?diff=unified&w=#diff-4e50dea4543eebdf47a29045f0fec793b706682fcb8fdd3013b8fb03f097b8b1) (and this is based on an older commit of @hpatro's PR). This really boils down to using `dictEntry` as an opaque structure as well and forcing access through common macros (SET/GET/GET_PTR) that perform common type translations. I think the changes are not that significant and if we could just do the first part (using `dictEntry` as an opaque structure) my concern will be addressed. Other refactoring can be addressed in future PRs.

hpatro committed:
Fix offset calculation for 32 bit

Signed-off-by: Harkrishn Patro <harkrisp@amazon.com>

hpatro committed:
Make clang format happy

Signed-off-by: Harkrishn Patro <harkrisp@amazon.com>

hpatro committed:
Update code comment

Signed-off-by: Harkrishn Patro <harkrisp@amazon.com>

hpatro commented:
The builds are passing for both 32 bit/64 bit now.

madolson commented on src/server.c:
```diff
@@ -464,14 +468,14 @@ dictType zsetDictType = {
-    dictObjectDestructor, /* val destructor */
-    dictResizeAllowed,    /* allow to resize */
-};
+dictType dbDictType = {dictSdsHash,          /* hash function */
+                       NULL,                 /* key dup */
+                       dictSdsKeyCompare,    /* key compare */
+                       NULL,                 /* key is embedded in the dictEntry and freed internally */
+                       dictObjectDestructor, /* val destructor */
+                       dictResizeAllowed,    /* allow to resize */
+                       .embedKey = dictSdsEmbedKey,
+                       .embedded_entry = 1};
 
```
Was the previous version giving format errors. I think we still generally prefer not introducing formatting changes if we don't have to.

hpatro replied:
Yeah.
https://github.com/valkey-io/valkey/actions/runs/9704277218/job/26784193065

madolson replied:
I think I fixed it.

madolson commented on src/server.h:
```diff
@@ -3558,6 +3558,7 @@ int dictSdsKeyCaseCompare(dict *d, const void *key1, const void *key2);
 void dictSdsDestructor(dict *d, void *val);
 void dictListDestructor(dict *d, void *val);
 void *dictSdsDup(dict *d, const void *key);
+size_t dictSdsKeyToBytes(unsigned char *buf, const void *key, unsigned char *header_size);
```
```suggestion
```

madolson commented on src/sds.c:
```diff
@@ -192,6 +192,25 @@ sds sdsdup(const sds s) {
+
+/* This method copies the sds `s` into `buf` which is the target character buffer. */
+size_t sdscopytobuffer(unsigned char *buf, size_t buf_len, sds s, uint8_t *hdr_size) {
+    size_t reqd_keylen = sdsminlen(s);
+    if (buf == NULL) {
+        return reqd_keylen;
+    }
+    assert(buf_len >= reqd_keylen);
+    memcpy(buf, sdsAllocPtr(s), reqd_keylen);
```
```suggestion
    size_t required_keylen = sdsminlen(s);
    if (buf == NULL) {
        return required_keylen;
    }
    assert(buf_len >= required_keylen);
    memcpy(buf, sdsAllocPtr(s), required_keylen);
```
I don't think we should use abbreviations unless they are consistently used.

zuiderkwast commented on src/dict.h:
```diff
@@ -80,8 +84,9 @@ typedef struct dictType {
-     * flag is set. */
+    /* If embedded_entry flag is set, it indicates that a copy of the key is created and the key is embedded
+     * as part of the dict entry. */
+    unsigned int embedded_entry : 1;
```
Some dependencies between these fields:

* If this flag is set, embedKey must also be set and keyDup and keyDestructor are forbidden?
* If this flag is not set, then embedKey must be NULL.

Correct? Can we add some debug asserts for that somewhere?

hpatro replied:
Correct. Added regular assert. I don't think it will be a lot of penalty. If you think otherwise, need to get in the helper debug assert methods to dict.

hpatro committed:
Update code comment and variable naming

Signed-off-by: Harkrishn Patro <harkrisp@amazon.com>

madolson commented:
I see 4/6 people directionally inclined, so going to throw on the directionally approved tag.

madolson commented on src/server.c:
```diff
@@ -464,14 +468,14 @@ dictType zsetDictType = {
-    dictObjectDestructor, /* val destructor */
-    dictResizeAllowed,    /* allow to resize */
-};
+dictType dbDictType = {dictSdsHash,          /* hash function */
+                       NULL,                 /* key dup */
+                       dictSdsKeyCompare,    /* key compare */
+                       NULL,                 /* key is embedded in the dictEntry and freed internally */
+                       dictObjectDestructor, /* val destructor */
+                       dictResizeAllowed,    /* allow to resize */
+                       .embedKey = dictSdsEmbedKey,
+                       .embedded_entry = 1};
```
```suggestion
dictType dbDictType = {
    dictSdsHash,          /* hash function */
    NULL,                 /* key dup */
    dictSdsKeyCompare,    /* key compare */
    NULL,                 /* key is embedded in the dictEntry and freed internally */
    dictObjectDestructor, /* val destructor */
    dictResizeAllowed,    /* allow to resize */
    .embedKey = dictSdsEmbedKey,
    .embedded_entry = 1,
};
```

Apparently makes clang happy? I don't get clang.

madolson committed:
Revert some formatting changes

Signed-off-by: Madelyn Olson <madelyneolson@gmail.com>

madolson commented on src/kvstore.h:
```diff
@@ -71,6 +70,12 @@ typedef dict *(kvstoreDictLUTDefragFunction)(dict *d);
+ * - If `dictType.embedded-entry` is 1, it clones the `key`.
+ * - Otherwise, it assumes ownership of the `key`.
+ * The caller must ensure the `key` is properly freed.
+ */
```
Can we put this on the C file instead of the header file? I'm okay if the header file defines everything, but feels weird here.

madolson commented on src/db.c:
```diff
@@ -241,15 +247,19 @@ int getKeySlot(sds key) {
+ * The kvstore handles `key` based on `dictType` during initialization:
+ * - If `dictType.embedded-entry` is 1, it clones the `key`.
+ * - Otherwise, it assumes ownership of the `key`.
+ * In this case a copy of `key` is made in kvstore, the caller must ensure the `key` is properly freed.
```
```suggestion
 * In this case a copy of `key` is copied in kvstore, the caller must ensure the `key` is properly freed.
```

madolson reviewed (approved):
LGTM, just some minor documentation changes that would make some stuff clearer.

zuiderkwast commented on src/db.c:
```diff
@@ -190,7 +190,14 @@ robj *lookupKeyWriteOrReply(client *c, robj *key, robj *reply) {
 }
 
-/* Add the key to the DB. It's up to the caller to increment the reference
+/* Add the key to the DB.
+ *
+ * The kvstore handles `key` based on `dictType` during initialization:
+ * - If `dictType.embedded-entry` is 1, it clones the `key`.
+ * - Otherwise, it assumes ownership of the `key`.
+ * In this case a copy of `key` is made in kvstore, the caller must ensure the `key` is properly freed.
```
This function (`dbAddInternal`) operates on the dbDictType, so always copies it, right?

zuiderkwast commented on src/db.c:
```diff
@@ -246,8 +253,13 @@ int getKeySlot(sds key) {
- * copy of the SDS string, otherwise 0 is returned, The caller should free the SDS string. */
+ * The function returns 1 if the key was added to the database, otherwise 0 is returned.
+ *
+ * The kvstore handles `key` based on `dictType` during initialization:
+ * - If `dictType.embedded-entry` is 1, it clones the `key`.
+ * - Otherwise, it assumes ownership of the `key`.
+ * In this case a copy of `key` is made in kvstore, the caller must ensure the `key` is properly freed.
+ */
 int dbAddRDBLoad(serverDb *db, sds key, robj *val) {
```
This function always uses dbDictType so it always copies.

zuiderkwast commented on src/kvstore.h:
```diff
@@ -70,6 +70,12 @@ typedef dict *(kvstoreDictLUTDefragFunction)(dict *d);
 void kvstoreDictLUTDefrag(kvstore *kvs, kvstoreDictLUTDefragFunction *defragfn);
 void *kvstoreDictFetchValue(kvstore *kvs, int didx, const void *key);
 dictEntry *kvstoreDictFind(kvstore *kvs, int didx, void *key);
+/*
+ * The kvstore handles `key` based on `dictType` during initialization:
+ * - If `dictType.embedded-entry` is 1, it clones the `key`.
+ * - Otherwise, it assumes ownership of the `key`.
+ * The caller must ensure the `key` is properly freed.
+ */
 dictEntry *kvstoreDictAddRaw(kvstore *kvs, int didx, void *key, dictEntry **existing);
```
Function docs are in the .c files in Valkey (not sure why, but it's consistently so).

I'm not completely sure about the kvstore abstraction, but I think we should mention in which cases each of these are used. The dictType is an implementation detail the caller of kvstore functions shouldn't need to worry about. For example say that for database keys (`db->keys`), it copies the key and for `db->expires` and sharded pubsub channels, it doesn't.

hpatro replied:
Moving some of the content to kvstoreDictCreate and adding the different usage(s) here.

zuiderkwast reviewed (commented):
Thanks for the added docs. I have a few follow-up comments on those.

hpatro committed:
Update code comment

Signed-off-by: Harkrishn Patro <harkrisp@amazon.com>

hpatro committed:
Add runtime check on dict type member dependency

Signed-off-by: Harkrishn Patro <harkrisp@amazon.com>

hpatro commented:
@zuiderkwast / @madolson Shall we ship it ? 🚀

zuiderkwast reviewed (approved):
Yes, LGTM.

bbarani commented:
> Yes, LGTM.

Awesome! Thanks all.

madolson commented:
Kicked off a full-run to stress this code a bit more before merging: https://github.com/valkey-io/valkey/actions/workflows/daily.yml

zvi-code commented on src/sds.c:
```diff
@@ -192,6 +192,25 @@ sds sdsdup(const sds s) {
+        return required_keylen;
+    }
+    assert(buf_len >= required_keylen);
+    memcpy(buf, sdsAllocPtr(s), required_keylen);
```
why do you need this double information about the header size? you can extract it from the "sds", right?

zuiderkwast replied:
An sds `s` is a pointer to where the string content starts, so it can be used as a C string. It does not point to the start of the allocation. The header is store before the char data in the same allocation, i.e. at `s[-1]`, `s[-2]`, etc. The header is backwards-encoded in some way, so the byte at `s[-1]` says which kind of header it is and how large the header is.

```
       sds-----.
               |
  allocation   v
  +-----------+----------------------+
  | hdr       |string contents    \0 |
  +-----------+----------------------+
```

When we store this embedded, we want to be able to restore the sds pointer, so we store the size of the header size in the first. When we restore the sds pointer, we can find it using S (the offset to the string contents).

```  
+-+-----------+----------------------+
|S| hdr       |string contents    \0 |
+-+-----------+----------------------+
```

zuiderkwast replied:
You mean the callers of this function can call `sdsHdrSize(s[-1])` themselves? I'm not sure if it's public.

zvi-code replied:
But we memcpy sdsAllocPtr and not sds, so you copy with the sds header, right?

zvi-code replied:
Yea, right agree. I am not so supportive of the desire to keep something as sds when it's not, but that's already in my overall comment, maybe I'll elaborate more there. Thanks

hpatro replied:
`key` is being used throughout the engine as a `sds` so changing that would be even more touchpoint. We could dynamically build it on the `dictGetKey` call but that would be additional penalty. Hence, storing it as a sds made the most sense.

judeng replied:
I have the same question, the low 3bit in sdshdr.flag have told us the header length, and exposing function `sdsHdrSize` will not lead to coupling between dict.c and sds.c, I don't see any harm in it.
```
                                                            
 sdshdr               sds                                   
   │                  │                                     
   │                  │                                     
   ▼─────┬─────┬─────┌▼───────────────────────────────┬──┐  
   │ len │ alloc     │                                │  │  
   │     │     │flags│                                │\0│  
   └─────┼─────┼─────└────────────────────────────────┴──┘  
                                                                                                                                                                                            
```

zvi-code commented:
Conceptually I feel it's not a good idea that we have 2 patterns: 1) use the internal encoding information of sds [pointer is odd] and 2) embed sds into some other non-native allocation buffer, potentially at arbitrary offset. I feel this has very big potential for very hard to track bugs

zvi-code commented:
> This change makes sense to me overall and tradeoffs are good.
> 
> 
> 
> My concern is with that conceptually it's not a good idea that we have 2 conflicting patterns: 
> 
> 
> 
> 1.  use the internal encoding information of sds [pointer is odd] 
> 
> 2. embed sds into some other non-native allocation buffer, potentially at arbitrary offset 
> 
> 
> 
> I feel this has very big potential for very hard to track bugs

Trying to arrange my thoughts on this. The above is just an example for a wider issue In the code design. IMO there are two types of 'sds' usages 1) A way to carry metadata about a string buffer through IO flow. 2) A compact way to encode string buffer metadata in memory with good locality and low memory overhead.

It is clear to me why use 'sds' for #2. For #1 I feel it could be wrong choice and this choice also has performance costs as it forces memory access to unpack the info many times in IO flow. For example sdsfree access the memory on free, even though it could have been avoided. This has high costs when memory access is a factor. 
I also think there is an alternative, to take an approach slightly similar to rust, have some struct 'runtimeSds' with sds as member and metadata, need to check if it can be passed to functions by value, to avoid the need to allocate it on the heap. Thoughts?

hpatro commented:
> > This change makes sense to me overall and tradeoffs are good.
> > My concern is with that conceptually it's not a good idea that we have 2 conflicting patterns:
> > 
> > 1. use the internal encoding information of sds [pointer is odd]
> > 2. embed sds into some other non-native allocation buffer, potentially at arbitrary offset
> > 
> > I feel this has very big potential for very hard to track bugs
> 
> Trying to arrange my thoughts on this. The above is just an example for a wider issue In the code design. IMO there are two types of 'sds' usages 1) A way to carry metadata about a string buffer through IO flow. 2) A compact way to encode string buffer metadata in memory with good locality and low memory overhead.
> 
> It is clear to me why use 'sds' for #2. For #1 I feel it could be wrong choice and this choice also has performance costs as it forces memory access to unpack the info many times in IO flow. For example sdsfree access the memory on free, even though it could have been avoided. This has high costs when memory access is a factor. I also think there is an alternative, to take an approach slightly similar to rust, have some struct 'runtimeSds' with sds as member and metadata, need to check if it can be passed to functions by value, to avoid the need to allocate it on the heap. Thoughts?

@zvi-code Thanks for sharing your thought. Would you mind filling a separate issue about this? Given we are planning to rehaul the hashtable implementation (#169) in 9.0. It will be good to capture some of these points.

madolson commented:
Actual test run I kicked off: https://github.com/valkey-io/valkey/actions/runs/9764931105

judeng commented:
This pr will bring us considerable cost benefits, thank you @hpatro . I have  doubt  about the benchmark test results, this PR should improve the cpu cache hit, but why hasn't the performance improved?

hpatro commented:
> This pr will bring us considerable cost benefits, thank you @hpatro . I have doubt about the benchmark test results, this PR should improve the cpu cache hit, but why hasn't the performance improved?

Yeah, I was also expecting more gain in throughput as well but saw a tiny gain. Anyway we were happy with the memory savings (without paying any cost). @judeng Do you have any recommendation to perform benchmarking to be able to observe the improvement in performance from CPU cache locality ?
